src/category_theory/groupoid/free_groupoid.lean

```diff
@@ -0,0 +1,198 @@
+/-
+Copyright (c) 2022 Rémi Bottinelli. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémi Bottinelli
+-/
+import category_theory.category.basic
+import category_theory.functor.basic
+import category_theory.groupoid
+import combinatorics.quiver.basic
+import combinatorics.quiver.connected_component
+import logic.relation
+import tactic.nth_rewrite
+import category_theory.path_category
+import category_theory.quotient
+
+/-!
+# Free groupoid on a quiver
+
+This file defines the free groupoid on a quiver, the lifting of a prefunctor to its unique
+extension as a functor from the free groupoid, and proves uniqueness of this extension.
+
+## Main results
+
+Given the type `V` and a quiver instance on `V`:
+
+- `free_groupoid V`: a type synonym for `V`.
+- `free_groupoid_groupoid`: the `groupoid` instance on `free_groupoid V`.
+- `lift`: the lifting of a prefunctor from `V` to `V'` where `V'` is a groupoid, to a functor.
+  `free_groupoid V ⥤ V'`.
+- `lift_spec` and `lift_unique`: the proofs that, respectively, `lift` indeed is a lifting
+  and is the unique one.
+
+## Implementation notes
+
+The free groupoid is first defined by symmetrifying the quiver, taking the induced path category
+and finally quotienting by the reducibility relation.
+
+-/
+
+
+open set classical function relation
+local attribute [instance] prop_decidable
+
+namespace category_theory
+namespace groupoid
+namespace free
+
+universes u v u' v'
+
+variables {V : Type u} [quiver.{v+1} V]
+
+/-- Shorthand for the "forward" arrow corresponding to `f` in `symmetrify V` -/
+abbreviation quiver.hom.to_pos {X Y : V} (f : X ⟶ Y) :
+  (quiver.symmetrify_quiver V).hom X Y := sum.inl f
+
+/-- Shorthand for the "backward" arrow corresponding to `f` in `symmetrify V` -/
+abbreviation quiver.hom.to_neg {X Y : V} (f : X ⟶ Y) :
+  (quiver.symmetrify_quiver V).hom Y X := sum.inr f
+
+/-- Shorthand for the "forward" arrow corresponding to `f` in `paths $ symmetrify V` -/
+abbreviation quiver.hom.to_pos_path {X Y : V} (f : X ⟶ Y) :
+  ((category_theory.paths.category_paths $ quiver.symmetrify V).hom X Y) := f.to_pos.to_path
+
+/-- Shorthand for the "forward" arrow corresponding to `f` in `paths $ symmetrify V` -/
+abbreviation quiver.hom.to_neg_path {X Y : V} (f : X ⟶ Y) :
+  ((category_theory.paths.category_paths $ quiver.symmetrify V).hom Y X) := f.to_neg.to_path
+
+/-- `p` and `q` are related if `p` is and `𝟙 X` and `q` is a back & forth -/
+def red_step : hom_rel $ paths $ quiver.symmetrify V :=
+λ X Y p q, ∃ (h : Y = X) (Z) (f : (quiver.symmetrify_quiver V).hom X Z),
+  (h.rec_on p = 𝟙 X) ∧ (h.rec_on q = f.to_path ≫ (quiver.reverse f).to_path)
```

Comment:
I would suggest using an inductive definition, which avoids the annoying equality `h` :
```lean
inductive red_step : hom_rel (paths (quiver.symmetrify V))
| step (X Z : quiver.symmetrify V) (f : X ⟶ Z) :
    red_step (𝟙 X) (f.to_path ≫ (quiver.reverse f).to_path)
```

Reply:
huh, that's quite a magical type… now using it, thanks!